### DISCUSSION TITLE
WebServer Methods Questions

### DISCUSSION THREAD

**Discussion by Xylopyrographer** [espressif/arduino-esp32#7656] [Q&A]
I'm looking to better understand what is going on with a few of the [WebServer](https://github.com/espressif/arduino-esp32/tree/master/libraries/WebServer) library methods.

Referring to the [WebUpdate](https://github.com/espressif/arduino-esp32/blob/master/libraries/WebServer/examples/WebUpdate/WebUpdate.ino) example:

Regarding the `send()` [methods](https://github.com/espressif/arduino-esp32/blob/41a83c4c02190c597a436686a0e58bcae34fcf08/libraries/WebServer/src/WebServer.h#L122):

1. What is the default header sent when using `send()`? (Assuming no calls to `sendHeader()` have been made previously.)

Regarding `sendHeader(const String& name, const String& value, bool first = false)` [method](https://github.com/espressif/arduino-esp32/blob/41a83c4c02190c597a436686a0e58bcae34fcf08/libraries/WebServer/src/WebServer.h#L135).

2. What is the effect on the header of using the `first` parameter? Does it change the order of elements assembled by previous `sendHeader()` calls?
    _Edit: In looking at  WebServer.cpp, confirming that `first` sets this header at the top of the response header string._

Regarding `on(const Uri &uri, HTTPMethod method, THandlerFunction fn, THandlerFunction ufn)` [here](https://github.com/espressif/arduino-esp32/blob/41a83c4c02190c597a436686a0e58bcae34fcf08/libraries/WebServer/src/WebServer.h#L90):

As used in [line 30](https://github.com/espressif/arduino-esp32/blob/41a83c4c02190c597a436686a0e58bcae34fcf08/libraries/WebServer/examples/WebUpdate/WebUpdate.ino#L30), it has 4 parameters.

4. For the `method` parameter, most examples use `HTTP_POST` or `HTTP_GET`. What do these expand to and where are they enumerated?

5. Trying to understand the 4th parameter, `ufn`. It seems to be the function called to handle file uploads. But under what conditions, or what triggers it to be called instead of the 3rd parameter, `fn`? 

   In the example, `fn` is called after `ufn` returns? but don't see what triggers that call, since the entire `on` statement is only executed if the endpoint is matched and is of type `HTTP_POST`.

6. Why prefer that method over using `onFileUpload(THandlerFunction fn)`? [here](https://github.com/espressif/arduino-esp32/blob/41a83c4c02190c597a436686a0e58bcae34fcf08/libraries/WebServer/src/WebServer.h#L94)

Looks like WebServer might be a candidate for refactoring as per #6759, but in the meantime any insight would be appreciated.

@igrr @SuGlider

**Comment by blaticslm** [comment]
These are function prototype of ```send()```
![图片](https://user-images.githubusercontent.com/47236078/211183962-24839703-3dd8-4146-9df7-05c9e3139b31.png)

As you can see, the ```String header``` is the header that need to send
![图片](https://user-images.githubusercontent.com/47236078/211183986-0ad0cbc5-6120-4ac0-b896-7e7e64afa4e9.png)

and then you can see the ```_prepareHeader``` function
```
void WebServer::_prepareHeader(String& response, int code, const char* content_type, size_t contentLength) {
    response = String(F("HTTP/1.")) + String(_currentVersion) + ' ';
    response += String(code);
    response += ' ';
    response += _responseCodeToString(code);
    response += "\r\n";

    using namespace mime;
    if (!content_type)
        content_type = mimeTable[html].mimeType;

    sendHeader(String(F("Content-Type")), String(FPSTR(content_type)), true);
    if (_contentLength == CONTENT_LENGTH_NOT_SET) {
        sendHeader(String(FPSTR(Content_Length)), String(contentLength));
    } else if (_contentLength != CONTENT_LENGTH_UNKNOWN) {
        sendHeader(String(FPSTR(Content_Length)), String(_contentLength));
    } else if(_contentLength == CONTENT_LENGTH_UNKNOWN && _currentVersion){ //HTTP/1.1 or above client
      //let's do chunked
      _chunked = true;
      sendHeader(String(F("Accept-Ranges")),String(F("none")));
      sendHeader(String(F("Transfer-Encoding")),String(F("chunked")));
    }
    if (_corsEnabled) {
        sendHeader(String(FPSTR("Access-Control-Allow-Origin")), String("*"));
	sendHeader(String(FPSTR("Access-Control-Allow-Methods")), String("*"));
	sendHeader(String(FPSTR("Access-Control-Allow-Headers")), String("*"));
    }
    sendHeader(String(F("Connection")), String(F("close")));

    response += _responseHeaders;
    response += "\r\n";
    _responseHeaders = "";
}
```

And I think those two functions above will answer your questions 1 and 2 .

**Comment by blaticslm** [comment]
For question 4, I would like to say if you are using ```Arduino 2.0.x``` or ```Platform IO```, you can hold left ```CTRL``` and click these two variables ```HTTP_POST```, ```HTTP_GET```. Arduino 2.0.x requires you to first verify and build so that the IntelliSense feature can be turned on. 

Question 5 may be solved by yourself if you can directly see the source code. For the part where you say how to trigger ```ufn``` and ```fn```. Based on my understanding, HTTP is like this: your mom calls you to eat dinner, and then you respond. 

So, assuming ```fn``` is the sending request function, and ```ufn``` is the hearing back function. You send the request to the server via ```fn```, and the server gives your response, handled by ```ufn```. 

I may not be correct, but this is what I have understood so far. Any mistakes please point them out because I am still learning. Thanks.
